### PR TITLE
[SMTChecker] Support modifiers

### LINKED ISSUE #5193: SMTChecker: support modifiers

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #5193

### CODE REVIEW

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -92,14 +92,20 @@ void SMTChecker::endVisit(VariableDeclaration const& _varDecl)
...
+	return false;
+}
+
 bool SMTChecker::visit(FunctionDefinition const& _function)
 {
-	if (!_function.modifiers().empty() || _function.isConstructor())
+	if (_function.isConstructor())
 		m_errorReporter.warning(
 			_function.location(),
 			"Assertion checker does not yet support constructors and functions with modifiers."
```
Comment (ekpyron) [inline libsolidity/formal/SMTChecker.cpp]:
```suggestion
			"Assertion checker does not yet support constructors."
```

**Review by ekpyron — COMMENTED** [review]
Looks OK apart from the minor comment, but I'm not entirely sure yet - need to have another look.

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -114,7 +120,55 @@ bool SMTChecker::visit(FunctionDefinition const& _function)
...
+	{
+		solAssert(m_modifierDepthStack.back() < int(function.modifiers().size()), "");
+		ASTPointer<ModifierInvocation> const& modifierInvocation = function.modifiers()[m_modifierDepthStack.back()];
+		solAssert(modifierInvocation, "");
+		modifierInvocation->accept(*this);
+		auto const& modifierDef = dynamic_cast<ModifierDefinition const&>(
+			*modifierInvocation->name()->annotation().referencedDeclaration
+		);
+		std::vector<ASTPointer<Expression>> const& modifierArguments =
+			modifierInvocation->arguments() ? *modifierInvocation->arguments() : std::vector<ASTPointer<Expression>>();
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: const ref to a temporary.

File: libsolidity/formal/SMTChecker.cpp [hunk 3]
```diff
@@ -114,7 +120,55 @@ bool SMTChecker::visit(FunctionDefinition const& _function)
...
+		vector<smt::Expression> modifierArgsExpr;
+		for (auto arg: modifierArguments)
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: You could just wrap this inside an if

File: test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+		_;
+	}
+
+	function f(uint x) m(x) m(2) m(s) public view {
+		assert(x > 0);
+		assert(s > 0);
+	}
+}
+// ----
+// Warning: (127-132): Condition is always true.
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol:18]: Does the rest of the warning make it clear that this comes from a call through f?

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add tests for:
 - modifiers that have local variables with the same names as local variables in the "calling" function
 - modifiers that call local functions (causing inlining) which might in turn invoke modifiers (even the same modifiers)
 - the same, but it causes recursion
 - test that shows that the modifier parameter receives a copy (not sure if this is obviously working, I just remember that this was a problem in my early implementation of modifiers)
 - test that `return` works correctly (jumps one level higher in the modifier stack, not out of the function altogether)

File: test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol (hunk 4 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol:18]: It doesn't... maybe we should add the call path as secondary location.

File: libsolidity/formal/SMTChecker.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/formal/SMTChecker.cpp]: Which is fine :-).

Comment (ekpyron) [reply libsolidity/formal/SMTChecker.cpp]: We had that discussion before, hadn't we? Constant references to temporaries will keep the temporary alive for the scope.

File: libsolidity/formal/SMTChecker.cpp (hunk 3 above)
Comment (ekpyron) [reply libsolidity/formal/SMTChecker.cpp]: Although I agree that this would be better in this case.

File: libsolidity/formal/SMTChecker.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp]: oh yes...

File: test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol (hunk 4 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/modifiers/modifier_parameters.sol:18]: Added in #6310

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
This crashes locally on macos:
```
/Users/alex/Projects/solidity/libsolidity/formal/SMTChecker.cpp:794:16: runtime error: reference binding to null pointer of type 'const dev::solidity::Expression'
```
It is this place:
```diff
diff --git a/libsolidity/formal/SMTChecker.cpp b/libsolidity/formal/SMTChecker.cpp
index 5085a7991..5cb291644 100644
--- a/libsolidity/formal/SMTChecker.cpp
+++ b/libsolidity/formal/SMTChecker.cpp
@@ -791,6 +791,7 @@ void SMTChecker::endVisit(Literal const& _literal)
 
 void SMTChecker::endVisit(Return const& _return)
 {
+       solAssert(_return.expression(), "");
        if (knownExpr(*_return.expression()))
        {
                auto returnParams = m_functionPath.back()->returnParameters();
```
In this test case:
```
/Users/alex/Projects/solidity/test/boostTest.cpp:114: error: in "smtCheckerTests/modifiers/modifier_return": Exception during extracted test: /Users/alex/Projects-Changes-ny/solidity/libsolidity/formal/SMTChecker.cpp(794): Throw in function virtual void dev::solidity::SMTChecker::endVisit(const dev::solidity::Return &)
Dynamic exception type: boost::exception_detail::clone_impl<langutil::InternalCompilerError>
std::exception::what: 
[dev::tag_comment*] = 
```

**Comment by leonardoalt** [comment]
@axic created https://github.com/ethereum/solidity/pull/6343 to fix the crash

**Pull request merged**